### PR TITLE
Card 610 better error handling

### PR DESCRIPTION
**Pull request by oarrietadotcms**

### CODE REVIEW

File: src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleResource.java [hunk 1]
```diff
@@ -237,7 +233,13 @@ private User getUser(@Context HttpServletRequest request) {
 
     private RestRule getRuleInternal(String ruleId, User user) {
         Rule input = getRule(ruleId, user);
-        return ruleTransform.appToRest(input);
+        RestRule restRule = ruleTransform.appToRest(input, user);
+
+        if(restRule == null) {
```
Comment (ggranum) [inline src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleResource.java]: the transform should never ever return null. if something changed we need to unchange it. 

#getRule(...) should throw a NotFoundException if the ruleId isn't present in the database. The transform.appToRest should throw any errors that prevent the method from returning a valid rule.

File: src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleTransform.java [hunk 2]
```diff
@@ -75,33 +77,50 @@ public Rule applyRestToApp(RestRule rest, Rule rule, User user) {
         return groups;
     }
 
-    public RestRule appToRest(Rule app) {
-        return toRest.apply(app);
-    }
+    public RestRule appToRest(Rule app, User user) {
+        RestRule restRule = toRest.apply(app);
```
Comment (ggranum) [inline src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleTransform.java]: We need to get out of the habit of returning null when bad things happen. It's a SUPER bad practice. ANd it leads to exactly this kind of defensive null checking that clutters up code is distracting.

File: src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleTransform.java [hunk 3]
```diff
@@ -75,33 +77,50 @@ public Rule applyRestToApp(RestRule rest, Rule rule, User user) {
...
+                    .shortCircuit(app.isShortCircuit())
+                    .priority(app.getPriority())
+                    .enabled(app.isEnabled())
+                    .conditionGroups(groups)
+                    .ruleActions(ruleActions)
+                    .build();
+
+        } catch (Exception e) {
+            String ruleName = UtilMethods.isSet(app.getName()) ? app.getName() : "N/A";
+            Logger.error(this, "Error parsing Rule named: " + ruleName + " to ReST", e);
```
Comment (ggranum) [inline src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleTransform.java]: throw new InternalServerException(.....). Or possibly throw a new custom exception type that extends that. RuleTransformException or similar.

File: src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleTransform.java (hunk 2 above)
Comment (dsilvam) [inline src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleTransform.java]: https://code.google.com/p/guava-libraries/wiki/UsingAndAvoidingNullExplained

File: src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleResource.java [hunk 4]
```diff
@@ -226,8 +219,19 @@ private User getUser(@Context HttpServletRequest request) {
...
+            List<RestRule> restRules = new ArrayList<>();
+
             List<Rule> rules = rulesAPI.getAllRulesByParent(host, user, false);
-            return rules.stream().map(ruleTransform.appToRestFn()).collect(Collectors.toList());
+            for (Rule rule : rules) {
+                try{
+                    restRules.add(ruleTransform.appToRest(rule, user));
+                } catch (Exception transEx) {
+                    String ruleName = UtilMethods.isSet(rule.getName()) ? rule.getName() : "N/A";
+                    Logger.error(this, "Error parsing Rule named: " + ruleName + " to ReST");
```
Comment (ggranum) [inline src/com/dotcms/rest/api/v1/sites/ruleengine/rules/RuleResource.java]: Make sure not to swallow the stack trace. Also, just let the rule print its own debug via the tostring method:
`Logger.error(this, ex, "Error transforming Rule: " + rule);`

**Pull request merged**